### PR TITLE
Integrated CWL support and Collections support (for both WDL and CWL)

### PR DESCRIPTION
**Pull request by AmandaBirmingham**
This PR ports and extends the functionality of q2cwl (https://github.com/qiime2/q2cwl ) into the q2dataflow framework.  Additionally, it adds support for the new Collection type in both WDL and CWL.

I realize this is a rather giant PR.  If there are guidelines/suggestions on breaking it up to make it more readable, or if we should talk through it live before the review, let me know.

Notes for reviewers:
* I was unable to use conda to create an environment containing qiime2 release 2023.7, q2dataflow, miniwdl, *and* cwltool; conda started throwing unsolvable-conflict errors. Mamba, however, worked like a charm.
* The Dockerfile used for the WDL testing for this branch specifically references this branch's name; this will have to be changed before integration with main.
* Support for Collection[Artifact] inputs in WDL has been implemented, but cannot be effectively tested due to the miniwdl emulator's propensity to move inputs around and thus break relative paths.  Per discussion on Campfire, tests for this class of inputs are thus skipped and return a message to this effect.
* The handling of multiple-type inputs in both CWL and WDL depends on treating all inputs as strings at collection and using parse_primitives to figure out their "real" type on consumption by Qiime2. This strategy *will fail* for edge cases like, say, wanting to input the *string* "1.5" rather than the float `1.5`.  Earlier discussion indicated we could live with this limitation, but we should definitely keep it in mind.

Question for reviewers:
1) This does not port the docker-based entry point in q2cwl, only the conda-based one.  Is the docker one still needed?  It wasn't clear to me what its use-case is.
2) What is needed to update the CI for this project to integrate the CWL tests?

### CODE REVIEW

**Comment by ebolyen** [comment]
Thanks @AmandaBirmingham, this is going to take me a bit to review. I will probably do it in multiple chunks over a few days. So I will report with a "finished reviewing" on the last one.

File: q2dataflow/languages/cwl/templaters/helpers.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    def outputs(self):
+        return {}
+
+
+class CwlInputCase(CwlParamCase):
+    def __init__(self, name, spec, arg=None, type_name=_cwl_file_type,
+                 is_optional=None, default=None, multiple=False):
+        super().__init__(name, spec, arg, type_name, is_optional, default)
+        self.multiple = multiple
+        self._is_file = self.spec.qiime_type.name != QIIME_COLLECTION_TYPE
```
Comment (ebolyen) [inline q2dataflow/languages/cwl/templaters/helpers.py]:
```suggestion
        self._is_file = (
            self.spec is None 
            or self.spec.qiime_type.name != QIIME_COLLECTION_TYPE
        )
```

File: q2dataflow/languages/cwl/templaters/helpers.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    def __init__(self, name, spec, arg=None, type_name=QIIME_STR_TYPE,
+                 is_optional=None, default=None):
+        super().__init__(name, spec, arg, type_name, is_optional, default)
+
+    def outputs(self):
+        out_binding_str = f"$(inputs.{self.name})"
+        type_suffix = "file"
+        cwl_type = _cwl_file_type
+
+        if self.spec.qiime_type.name == 'Collection':
```
Comment (ebolyen) [inline q2dataflow/languages/cwl/templaters/helpers.py]:
```suggestion
        if self.spec is not None and self.spec.qiime_type.name == 'Collection':
```

File: q2dataflow/languages/wdl/templaters/helpers.py [hunk 3]
```diff
@@ -89,32 +100,52 @@ def _make_array_inputs(name, array_inner_type, is_optional, default,
...
+    dataflow_prefix = q2wdl_prefix
+    _dataflow_keywords = _wdl_keywords_v1
+
+    def __init__(self, name, spec, arg=None, type_name=None,
+                 is_optional=None, default=None):
+        super().__init__(
+            name, spec, arg, type_name, is_optional, default)
+
+        self._is_collection = \
+            self.spec.qiime_type.name == QIIME_COLLECTION_TYPE
```
Comment (ebolyen) [inline q2dataflow/languages/wdl/templaters/helpers.py]:
```suggestion
        self._is_collection = (
            self.spec is not None
            and self.spec.qiime_type.name == QIIME_COLLECTION_TYPE
        )
```

**Review by ebolyen — CHANGES_REQUESTED** [review]
Some immediate notes: the builtins don't quite work because the `spec` is initialized to `None` for the (manually created) case instances.

I like how the WdlParamCase base-class handles the collection question all at once, and I think we can do the same with CWL. I pointed out some minimal changes which technically fix the issue, but there's no obligation to actually use them.

Would there be utility in bumping this logic up into the root ParamCase?

**Comment by Oddant1** [comment]
@AmandaBirmingham Is this PR still in development? I've been meaning to look at it, but it's huge and I don't feel I know enough to do a super comprehensive review in addition to the unresolved comments by @ebolyen.

**Comment by AmandaBirmingham** [comment]
@Oddant1 Yes, my apologies, you will see a commit tomorrow!

**Comment by Oddant1** [comment]
@AmandaBirmingham All good! Just checking in.

**Comment by AmandaBirmingham** [comment]
@ebolyen , I think you set up the CI for the wdl tests (which are hideous because they have to run through docker).  I have made the changes above, which (after rebuilding my docker image and forcing a `docker swarm leave`) allows me to successfully create templates for builtins on my local machine and also to run the WDL mystery stew tests.  However, on github the WDL unit tests are failing at the first test, saying `"docker image not found: testq2dataflow:latest"` ... do you have any idea what might lead to this issue in the CI environment?  (I mean, presumably the lack of a docker image is leading to this issue ... but I'm wondering why no image ;)

**Comment by AmandaBirmingham** [comment]
@ebolyen @Oddant1 Hi, Folks--a ping to see if either of you can help me figure out what the issue with the CI docker image generation/access is here?  The unit tests won't run (and pass and get this PR merged, I hope ;) because `miniwdl` reports a `"docker image not found: testq2dataflow:latest"` error.   AFAICT (which is not very far ...) it looks like the docker image is being created without error earlier in the CI, so I am unclear why its image isn't available at the later step.  I think the docker functionality in the github workflow was originally the magic of @ebolyen ... do you have any insight into what might be going weird here?

**Comment by ebolyen** [comment]
Hey @AmandaBirmingham, unfortunately, I don't have any good insight here yet. But I will be overhauling some of our docker builds for our releases here in the next week, and I bet I can come up with a smarter way to handle this once I've tested a few things in that context.

**Comment by ebolyen** [comment]
Tests are actually running!

File: .github/workflows/ci_wdl.yml [hunk 4]
```diff
@@ -36,18 +37,18 @@ jobs:
...
-          environment-file: qiime2-latest-py38-linux-conda.yml
+          environment-file: qiime2-amplicon-ubuntu-latest-conda.yml
           auto-activate-base: false
 
       - name: Install q2dataflow
         shell: bash -el {0}
         run: |
-          pip install -e .
+          pip install .
           pip install miniwdl
```
Comment (ebolyen) [inline .github/workflows/ci_wdl.yml]:
```suggestion
          pip install miniwdl cwlref-runner
```

File: q2dataflow/core/signature_converter/util.py [hunk 5]
```diff
@@ -28,16 +28,22 @@ def get_mystery_stew(desired_filters=None):
...
-
-    pm.add_plugin(test_plugin)
+    try:
+        pm = sdk.PluginManager.reuse_existing()
+    except sdk.UninitializedPluginManagerError:
+        pm = sdk.PluginManager(add_plugins=False)
+
+    try:
+        pm.get_plugin(id='mystery_stew')
+    except KeyError:
```
Comment (ebolyen) [inline q2dataflow/core/signature_converter/util.py:38]: duh... of course, thanks :)

I was just so baffled how the pluginmanager was already instantiated (turned out to be this: https://github.com/qiime2/q2cli/issues/334)

**Comment by ebolyen** [comment]
```
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/qiime2/bin/q2dataflow", line 5, in <module>
    from q2dataflow.__main__ import root
  File "/usr/share/miniconda/envs/qiime2/lib/python3.8/site-packages/q2dataflow/__main__.py", line 13, in <module>
    from q2dataflow.languages.wdl.templaters.helpers import \
ModuleNotFoundError: No module named 'q2dataflow.languages'
```
(From the last run)
This looks like a circular import not having a finalized module to me

**Comment by ebolyen** [comment]
:tada: :tada: :tada:

**Comment by AmandaBirmingham** [comment]
@ebolyen  All WDL and CWL tests pass (except for the ones intentionally skipped, of course :) :rocket:

**Review by ebolyen — APPROVED** [review]
Thanks so much @AmandaBirmingham!

**Comment by AmandaBirmingham** [comment]
Yaay! Talk about a team effort! :raised_hands:

**Pull request merged**